python/cuml/feature_extraction/text.py

```diff
@@ -0,0 +1,603 @@
+            return list(ENGLISH_STOP_WORDS)
+        elif isinstance(self.stop_words, str):
+            raise ValueError("not a built-in stop list: %s" % self.stop_words)
+        elif self.stop_words is None:
+            return None
+        else:  # assume it's a collection
+            return list(self.stop_words)
+
+    def get_char_ngrams(self, n, str_series, doc_id_sr):
+        """
+        Handles ngram generation for characters analyzers.
+
+        When analyzer is 'char_wb', we generate ngrams within word boundaries,
+        meaning we need to first tokenize and pad each token with a delimiter.
+        """
+        if self.analyzer == 'char':
+            token_count = str_series.str.len()
+            ngram_sr = str_series.str.character_ngrams(n=n)
+            ngram_count = None
+        else:
+            # here analyzer is 'char_wb'
+            tokens = str_series.str.tokenize(self.delimiter)
+
+            padding = Series(self.delimiter).repeat(len(tokens))
+            tokens = padding.str.cat(tokens.str.cat(padding))
+            tokens = tokens.reset_index(drop=True)
+
+            ngram_sr = tokens.str.character_ngrams(n=n)
+
+            token_count = str_series.str.token_count(self.delimiter)
+            doc_id_df = cudf.DataFrame({
+                'doc_id': doc_id_sr.repeat(token_count).reset_index(drop=True),
+                # formula to count ngrams given number of letters per token:
+                'ngram_count': tokens.str.len() - (n - 1)
+            })
+            ngram_count = doc_id_df.groupby('doc_id').sum()['ngram_count']
+
+        return ngram_sr, ngram_count, token_count
+
+    def get_ngrams(self, str_series, n, doc_id_sr):
+        """
+        This returns the ngrams for the string series
+        Parameters
+        ----------
+        str_series : (cudf.Series)
+            String series to tokenize
+        n : int
+            Gram level to get (1 for unigram, 2 for bigram etc)
+        doc_id_sr : cudf.Series
+            Int series containing documents ids
+        """
+        ngram_count = None
+        if self.analyzer == 'word':
+            token_count_sr = str_series.str.token_count(self.delimiter)
+            ngram_sr = str_series.str.ngrams_tokenize(n=n, separator=" ",
+                                                      delimiter=self.delimiter)
+        else:
+            if n != 1:
+                ngram_sr, ngram_count, token_count_sr = self.get_char_ngrams(
+                    n, str_series, doc_id_sr
+                )
+            else:
+                token_count_sr = str_series.str.len()
+                ngram_sr = str_series.str.character_tokenize()
+
+        not_empty_docs = token_count_sr > 0
+        token_count_sr = token_count_sr[not_empty_docs]
+        doc_id_sr = doc_id_sr[not_empty_docs]
+
+        # formula to count ngrams given number of tokens x per doc: `x-(n-1)`
+        if ngram_count is None:
+            ngram_count = token_count_sr - (n - 1)
```

Comment:
oh, I see why n_gram count was none above... kind of a confusing flow. Maybe we just compute it directly in the earlier function?

Reply:
Good point, refactored this part, hopefully it's clearer now.